pandaserver/dataservice/closer.py

```diff
@@ -0,0 +1,469 @@
+        """
+        closer_plugin_class = panda_config.getPlugin("closer_plugins", self.job.VO)
+        if closer_plugin_class is None and self.job.VO == "atlas":
+            # use ATLAS plugin for ATLAS
+            from pandaserver.dataservice.closer_atlas_plugin import (
+                CloserAtlasPlugin,
+            )
+
+            closer_plugin_class = CloserAtlasPlugin
+        if closer_plugin_class is not None:
+            closer_plugin = closer_plugin_class(self.job, final_status_dataset, _logger)
+            closer_plugin.execute()
+
+    def start_notifier(self, disable_notifier: bool) -> None:
+        """
+        Start the notifier.
+
+        Args:
+            disable_notifier (bool): The flag indicating if the notifier should be disabled.
+        """
+
+        tmp_log = LogWrapper(_logger,
+                             f"start_notifier-{datetime.datetime.now(datetime.timezone.utc).replace(tzinfo=None).isoformat('/')}")
+
+        # don't send email for merge jobs
+        if (not disable_notifier) and self.job.processingType not in ["merge", "unmerge"]:
+            use_notifier = True
+            summary_info = {}
+            # check all jobDefIDs in jobsetID
+            if self.job.jobsetID not in [0, None, "NULL"]:
+                (
+                    use_notifier,
+                    summary_info,
+                ) = self.task_buffer.checkDatasetStatusForNotifier(
+                    self.job.jobsetID,
+                    self.job.jobDefinitionID,
+                    self.job.prodUserName,
+                )
+                tmp_log.debug(f"use_notifier:{use_notifier}")
+            if use_notifier:
+                tmp_log.debug("start Notifier")
+                notifier_thread = Notifier.Notifier(
+                    self.task_buffer,
+                    self.job,
+                    self.destination_dispatch_blocks,
+                    summary_info,
+                )
+                notifier_thread.run()
+                tmp_log.debug("end Notifier")
+
+    def close_user_datasets(self, dataset, final_status: str):
+        """
+        Close user datasets
+
+        Args:
+            dataset: Dataset.
+            final_status (str): Final status.
+        """
+        tmp_log = LogWrapper(_logger,
+                             f"close_user_datasets-{datetime.datetime.now(datetime.timezone.utc).replace(tzinfo=None).isoformat('/')}")
+
+        if (
+                self.job.prodSourceLabel in ["user"]
+                and self.job.destinationDBlock.endswith("/")
+                and (dataset.name.startswith("user") or dataset.name.startswith("group"))
+        ):
+            # get top-level user dataset
+            top_user_dataset_name = re.sub("_sub\d+$", "", dataset.name)
+            # update if it is the first attempt
+            if top_user_dataset_name != dataset.name and top_user_dataset_name not in self.top_user_dataset_list and self.job.lockedby != "jedi":
```

Comment:
Check if possible to delete

Reply:
Probably most of stuff except for starting activator in close_user_datasets() can be deleted since only HC jobs are relevant but they don't want to close their datasets.